### PR TITLE
Fix html code with unicode content

### LINKED ISSUE #9491: Reproducing <!doctype html> on every save of the email
[//]: # ( Invisible comment: 
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Before you create the issue:
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Search for similar report among other reported issues.
Learn how to troubleshoot at https://www.mautic.org/docs/en/tips/troubleshooting.html
Use drag&drop to attach images or other files )

## Bug Description
The more details the better...

| Q   | A
| --- | ---
| Mautic version | 3.2.1
| PHP version | 7.3.23
| Browser | Google Chrome

### Steps to reproduce
1. Open code mode email
2. Save the email
3. Every save will add exactly 1 more <!doctype html> tag to the begginig of the email
4. You will see the reproducing <!doctype html> tags in the beginning of the email.
 
### Log errors

No logs, here is the screenshot.

ScreenShot_20201105224203

The issue is appeared with Mauitc 3.1.x and still exists with Mauitc 3.2.1

Pull https://github.com/mautic/mautic/pull/669 handled the opposite of this issue.

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

### PR DESCRIPTION
**Pull request by kuzmany**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | 4.0
| Bug fix?                               | yes
| New feature?                           | no
| Deprecations?                          | no
| BC breaks?                             | no
| Automated tests included?              | yes
| Related user documentation PR URL      | 
| Related developer documentation PR URL | 
| Issue(s) addressed                     | 

#### Description:
This is PR to resolve https://github.com/mautic/mautic/issues/9491. More context what is issue https://github.com/mautic/mautic/issues/9491#issuecomment-748581544 
This issue bring this PR https://github.com/mautic/mautic/pull/9108 by @mabumusa1 - maybe you can help with it.
I've added unit test case, what we need resolve.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create landing page with legacy builder
3. Add         ` <a href="https://m3.mautibox.com/3.x/media/images/testá.png">test with unicode</a>`
4. Apply in builder 3 times
5. Go to page 
6. Open source of page
7. You should see duplicate `<!DOCTYPE html>` tag in the header of page

image

8. Apply patch, create page from scratch and repeat all steps again
9.  `<!DOCTYPE html>` should be in the header just once

<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/9516"></a>

### CODE REVIEW

**Review by npracht — DISMISSED** [review]
I confirm it works properly !!

**Review by trianity — DISMISSED** [review]
OK, this PR works and fix the issue.

File: app/bundles/CoreBundle/Helper/InputHelper.php [hunk 1]
```diff
@@ -443,16 +443,13 @@ function ($matches) {
...
             $hasUnicode = strlen($value) != strlen(utf8_decode($value));
 
-            // Encode the incoming value before cleaning, it convert unicode to encoded strings
-            $value = $hasUnicode ? rawurlencode($value) : $value;
-
-            $value = self::getFilter(true)->clean($value, 'html');
+            $value = self::getFilter(true)->clean($value, $hasUnicode ? 'raw' : 'html');
 
             // After cleaning encode the value
             $value = $hasUnicode ? rawurldecode($value) : $value;
```
Comment (MarketSmart) [inline app/bundles/CoreBundle/Helper/InputHelper.php:464]: @kuzmany Should this be removed as well since its been removed above.

Comment (kuzmany) [reply app/bundles/CoreBundle/Helper/InputHelper.php:464]: Don't understand question

Comment (trianity) [reply app/bundles/CoreBundle/Helper/InputHelper.php:464]: @kuzmany If I understand well, then @MarketSmart mentioned that these too lines 448, 449 in the white sections are unnecessary:
// After cleaning encode the value
$value = $hasUnicode ? rawurldecode($value) : $value;

Comment (kuzmany) [reply app/bundles/CoreBundle/Helper/InputHelper.php:464]: It's diff, red lines are removed and green lines are added.

Comment (trianity) [reply app/bundles/CoreBundle/Helper/InputHelper.php:464]: @kuzmany Yes, you are right. @MarketSmart - by my opinion - mentioned line 447 (red) and line (452) white which are the same. The line  "$value = $hasUnicode ? rawurldecode($value) : $value;" was duplicated in the original code. It really needs or typo?

**Comment by volha-pivavarchyk** [comment]
@kuzmany Could you please rebase this PR. We need this changes

**Review by jos0405 — DISMISSED** [review]
Yep, this works. Can you plz rebase so it can be merged after that?

**Comment by kuzmany** [comment]
@volha-pivavarchyk @jos0405 rebased

**Comment by escopecz** [comment]
@kuzmany is there any reason this was rebased on 4.x which is the feature branch and not on 4.4 which is the bug fix branch?

**Comment by RCheesley** [comment]
This PR requires a rebase so it is not mergeable at this time, adding the relevant labels but let's get it merged asap :)

**Comment by adiux** [comment]
@kuzmany https://github.com/mautic/mautic/pull/9516#issuecomment-1230277626

**Comment by kuzmany** [comment]
Rebased on 4.4

**Comment by escopecz** [comment]
Thank you Zdeno! I learned my lesson with the last release. The 4.4 branch cannot be merged into 5.x branch anymore so I had to cherry-pick all commits from the release to the 5.x branch one by one. Please create a duplicate PR against 5.x for this PR. It must be merged to 5.x first and then we can merge this PR into 4.4 to keep both versions up to date.

**Review by jos0405 — APPROVED** [review]
This still works. In case you can merge.

**Comment by trianity** [comment]
The issue is still exist in Mautic v.4.4.5 , so it would be good if the PR would accepted and merged.

**Review by mabumusa1 — APPROVED** [review]
The changes look good to me, I think this PR is good to go.

**Pull request merged**